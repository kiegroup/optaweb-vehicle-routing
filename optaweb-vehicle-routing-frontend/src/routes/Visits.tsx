/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Text, TextContent, TextVariants } from '@patternfly/react-core';
import LocationList from 'components/LocationList';
import * as React from 'react';
import { connect } from 'react-redux';
import { IAppState } from 'store/configStore';
import { demoOperations } from 'store/demo';
import { routeOperations, routeSelectors } from 'store/route';
import { IRoute } from 'store/route/types';

interface IStateProps {
  route: IRoute;
  domicileId: number;
  isDemoLoading: boolean;
}

const mapStateToProps = ({ route, demo }: IAppState): IStateProps => ({
  domicileId: routeSelectors.getDomicileId(route),
  isDemoLoading: demo.isLoading,
  route,
});

export interface IDispatchProps {
  removeHandler: typeof routeOperations.deleteLocation;
  loadHandler: typeof demoOperations.loadDemo;
  clearHandler: typeof routeOperations.clearRoute;
  addHandler: typeof routeOperations.addLocation;
}

const mapDispatchToProps: IDispatchProps = {
  addHandler: routeOperations.addLocation,
  clearHandler: routeOperations.clearRoute,
  loadHandler: demoOperations.loadDemo,
  removeHandler: routeOperations.deleteLocation,
};

type IProps = IStateProps & IDispatchProps;

class Visits extends React.Component<IProps> {

  constructor(props: IProps) {
    super(props);
  }

  render(): React.ReactNode {
    const {
      route,
      domicileId,
      removeHandler,
      loadHandler,
      clearHandler,
      isDemoLoading,
    } = this.props;
    return (
      <React.Fragment>
        <TextContent>
          <Text component={TextVariants.h1}>Visits ({route.locations.length})</Text>
        </TextContent>
        {/* TODO do not show depots */}
        <LocationList
          removeHandler={removeHandler}
          selectHandler={() => undefined}
          loadHandler={loadHandler}
          clearHandler={clearHandler}
          route={route}
          domicileId={domicileId}
          isDemoLoading={isDemoLoading}
        />
      </React.Fragment>
    );
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Visits);
