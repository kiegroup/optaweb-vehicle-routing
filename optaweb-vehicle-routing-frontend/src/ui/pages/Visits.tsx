/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Text, TextContent, TextVariants } from '@patternfly/react-core';
import * as React from 'react';
import { connect } from 'react-redux';
import { routeOperations } from 'store/route';
import { Location } from 'store/route/types';
import { AppState } from 'store/types';
import LocationList from 'ui/components/LocationList';

interface StateProps {
  depot: Location | null;
  visits: Location[];
}

const mapStateToProps = ({ plan }: AppState): StateProps => ({
  depot: plan.depot,
  visits: plan.visits,
});

export interface DispatchProps {
  addHandler: typeof routeOperations.addLocation;
  removeHandler: typeof routeOperations.deleteLocation;
}

const mapDispatchToProps: DispatchProps = {
  addHandler: routeOperations.addLocation,
  removeHandler: routeOperations.deleteLocation,
};

export type Props = StateProps & DispatchProps;

export const Visits: React.FC<Props> = ({
  depot,
  visits,
  removeHandler,
}: Props) => (
  <>
    <TextContent>
      <Text component={TextVariants.h1}>{`Visits (${visits.length})`}</Text>
    </TextContent>
    {/* TODO do not show depots */}
    <LocationList
      removeHandler={removeHandler}
      selectHandler={() => undefined}
      depot={depot}
      visits={visits}
    />
  </>
);

export default connect(mapStateToProps, mapDispatchToProps)(Visits);
