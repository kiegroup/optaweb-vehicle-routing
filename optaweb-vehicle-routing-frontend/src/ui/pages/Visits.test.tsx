/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Props, Visits } from './Visits';

describe('Visits page', () => {
  it('should render correctly with no visits', () => {
    const visits = shallow(<Visits {...noVisits} />);
    expect(toJson(visits)).toMatchSnapshot();
  });

  it('should render correctly with a few visits', () => {
    const visits = shallow(<Visits {...twoVisits} />);
    expect(toJson(visits)).toMatchSnapshot();
  });
});

const noVisits: Props = {
  addHandler: jest.fn(),
  removeHandler: jest.fn(),

  depot: null,
  visits: [],
};

const twoVisits: Props = {
  addHandler: jest.fn(),
  removeHandler: jest.fn(),

  depot: {
    id: 1,
    lat: 1.345678,
    lng: 1.345678,
  },

  visits: [{
    id: 2,
    lat: 2.345678,
    lng: 2.345678,
  }, {
    id: 3,
    lat: 3.676111,
    lng: 3.568333,
  }],
};
