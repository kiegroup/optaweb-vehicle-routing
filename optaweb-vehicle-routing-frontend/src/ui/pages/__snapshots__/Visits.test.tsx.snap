// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visits page should render correctly with a few visits 1`] = `
<Fragment>
  <TextContent>
    <Text
      component="h1"
    >
      Visits (2)
    </Text>
  </TextContent>
  <LocationList
    depot={
      Object {
        "id": 1,
        "lat": 1.345678,
        "lng": 1.345678,
      }
    }
    removeHandler={[MockFunction]}
    selectHandler={[Function]}
    visits={
      Array [
        Object {
          "id": 2,
          "lat": 2.345678,
          "lng": 2.345678,
        },
        Object {
          "id": 3,
          "lat": 3.676111,
          "lng": 3.568333,
        },
      ]
    }
  />
</Fragment>
`;

exports[`Visits page should render correctly with no visits 1`] = `
<Fragment>
  <TextContent>
    <Text
      component="h1"
    >
      Visits (0)
    </Text>
  </TextContent>
  <LocationList
    depot={null}
    removeHandler={[MockFunction]}
    selectHandler={[Function]}
    visits={Array []}
  />
</Fragment>
`;
